Adjust ThreadListStackTracesTest for checkReentrantLock

singh264 opened the pull request:
Adjust ThreadListStackTracesTest for checkReentrantLock
to be ready when the virtual thread state is WAITING after
the reentrant lock parks the virtual thread. When the
virtual thread is mounted, the virtual thread state, which
is obtained from the carrier thread, can transition
between WAITING and RUNNABLE before the virtual thread is
parked on the lock.

Issue: https://github.com/eclipse-openj9/openj9/issues/18089
Signed-off-by: Amarpreet Singh <Amarpreet.A.Singh@ibm.com>

singh264 commented:
@babsingh requesting review

babsingh commented on test/hotspot/jtreg/serviceability/jvmti/vthread/ThreadListStackTracesTest/ThreadListStackTracesTest.java:
```diff
@@ -97,6 +107,10 @@ private static void checkReentrantLock() throws InterruptedException {
         Thread vt = Thread.ofVirtual().name(name).start(task);
+        // wait until vt is waiting to acquire the lock
+        while (!reentrantLock.hasQueuedThread(vt)) {
+            sleep(1);
```
```suggestion
            TestTask.sleep(1);
```

babsingh commented on test/hotspot/jtreg/serviceability/jvmti/vthread/ThreadListStackTracesTest/ThreadListStackTracesTest.java:
```diff
@@ -86,6 +88,14 @@ public class ThreadListStackTracesTest {
     static void log(String msg) { System.out.println(msg); }
     static void failed(String msg) { throw new RuntimeException(msg); }
 
+    static void sleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interruption in ThreadListStackTracesTest.sleep: \n\t" + e);
+        }
+    }
+
```
`TestTask.sleep` can be reused.
```suggestion
```

babsingh commented on test/hotspot/jtreg/serviceability/jvmti/vthread/ThreadListStackTracesTest/ThreadListStackTracesTest.java:
```diff
@@ -45,6 +45,8 @@ static void sleep(long millis) {
     }
 
     public void ensureReady(Thread vt, Thread.State expState) {
+        // sleep for 1 second to allow the thread to reach the expected state
```
in the comment, mention about the timing gap where the state can fluctuate

babsingh commented:
In the commit message, the flow is better if `Thread.State.` is removed i.e. `Thread.State.WAITING -> WAITING`.

babsingh commented:
The test is flawed due to the timing gap. This change will reduce the frequency of the failure. In the worst case, the `checkReentrantLock` sub-test can be disabled via commenting if the failure still occurs due to the timing gap.

How less frequent is the failure with this PR? Does it never occur?

singh264 commented:
> How less frequent is the failure with this PR? Does it never occur?

The test does not fail with this PR when it is run 2500 times.

The test fails approximately once every 100 times without this PR.

babsingh commented on test/hotspot/jtreg/serviceability/jvmti/vthread/ThreadListStackTracesTest/ThreadListStackTracesTest.java:
```diff
@@ -45,6 +45,13 @@ static void sleep(long millis) {
     }
 
```
On line 23, IBM copyright needs to be added. See https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/60/files for an example.

```
/*
 * ===========================================================================
 * (c) Copyright IBM Corp. 2023, 2023 All Rights Reserved
 * ===========================================================================
 */
```

babsingh commented on test/hotspot/jtreg/serviceability/jvmti/vthread/ThreadListStackTracesTest/ThreadListStackTracesTest.java:
```diff
@@ -97,6 +104,10 @@ private static void checkReentrantLock() throws InterruptedException {
         String name = "ReentrantLockTestTask";
         TestTask task = new ReentrantLockTestTask();
         Thread vt = Thread.ofVirtual().name(name).start(task);
+        // wait until vt is waiting to acquire the lock
```
```suggestion
        // Wait until vt is waiting to acquire the lock.
```

babsingh reviewed (commented):
Also, squash the commits.

singh264 commented:
> Also, squash the commits.

@babsingh I have squashed the commits, updated the commit message, and addressed the new comments.

Re-requesting your review.

babsingh reviewed (approved):
300x grinder passed: https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/36164.

keithc-ca commented on test/hotspot/jtreg/serviceability/jvmti/vthread/ThreadListStackTracesTest/ThreadListStackTracesTest.java:
```diff
@@ -45,6 +45,13 @@ static void sleep(long millis) {
+         * WAITING and RUNNABLE before the reentrant lock parks the virtual thread
+         */
+        sleep(1000);
         // wait while the thread is not ready or thread state is unexpected
```
The existing comment from upstream should have been left alone; the change does nothing but increase the likelihood of future merge conflicts. Same comment for https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/693.

babsingh replied:
We will be more careful in the future. I can revert that line. Will reverting that line now prevent future merge conflicts?

keithc-ca replied:
It depends how close the next change from upstream is. It may or may not conflict, but in general we should avoid unnecessary changes to code from upstream.

babsingh replied:
Ack on the unnecessary changes. The risk for a merge conflict seems pretty low for that one-line. So, I feel that no preventive actions are needed.

keithc-ca replied:
Agreed. It was just a suggestion aimed at avoiding unwanted differences relative to upstream. When reasonable, I think we should look at _any_ change we make to openjdk code as an opportunity to review *all* differences as compared with upstream.

tajila commented:
@singh264 Need to double deliver to 0.42

singh264 commented:
> @singh264 Need to double deliver to 0.42

@tajila I have created a PR to merge the changes to branch `v0.42.0-release` https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/70.

babsingh commented:
- These changes are also needed for 0.43 since 0.43 is a Java 21 refresh.
- Milestone details: https://github.com/eclipse-openj9/openj9/milestones.
- JDK21 v0.43.0-release: https://github.com/ibmruntimes/openj9-openjdk-jdk21/tree/v0.43.0-release
